scaaml/capture/aes/capture_runner.py

```diff
@@ -0,0 +1,159 @@
+          capture_runner.capture()
+        """
+        self._crypto_algorithms = crypto_algorithms
+        self._communication = communication
+        self._control = control
+        self._scope = scope
+        self._dataset = dataset
+
+    def capture_trace(self, key: bytearray, plaintext: bytearray):
+        """Try to capture a single trace.
+
+        Args:
+          key: The key to use.
+          plaintext: The plaintext to be encrypted.
+
+        Returns: The captured trace. None if the capture failed.
+          See cw documentation for description of the trace.
+
+        Raises: Warning, OSError
+        """
+        # Try to capture trace.
+        return cw.capture_trace(scope=self._scope.scope,
+                                target=self._communication.target,
+                                plaintext=plaintext,
+                                key=key)
+
+    def get_attack_points_and_measurement(self, key: bytearray,
+                                          plaintext: bytearray,
+                                          crypto_alg: SCryptoAlgorithm):
+        """Get attack points and measurement. Repeat capture if necessary.
+        Raises if hardware fails.
+
+        Args:
+          key: The key to use.
+          plaintext: The plaintext to be encrypted.
+          crypto_alg: The object used to get attack points.
+
+        Returns: Attack points and physical measurement.  See cw documentation
+          for description of the trace.
+
+        Raises:
+          Warning
+          OSError
+          AssertionError When the textin of the cw trace is different from the
+            plaintext.
+        """
+        while True:  # Make sure to capture the trace.
+            trace = self.capture_trace(plaintext=plaintext, key=key)
+            if trace:
+                assert trace.textin == plaintext
+                attack_points = crypto_alg.attack_points(plaintext=plaintext,
+                                                         key=key)
+                measurement = {
+                    "trace": trace.wave,
+                }
+                return attack_points, measurement
+
+    def _stabilize_capture(self, crypto_alg: SCryptoAlgorithm):
+        """Stabilize the capture by capturing a few traces.
+
+        Args:
+          crypto_alg: The object used to get stabilization attack points.
+        """
+        skti = crypto_alg.get_stabilization_kti()
+        key, text = next(skti)
+        cur_pt = bytearray(text)
+        cur_key = bytearray(key)
+        try:
+            self.capture_trace(plaintext=cur_pt, key=cur_key)
+        except (Warning, OSError):
+            pass
+        # Stabilize the capture
+        for _ in range(10):
+            self.capture_trace(plaintext=cur_pt, key=cur_key)
+
+    def capture(self):
+        """Start (or resume) and finish the capture."""
+        for crypto_alg in self._crypto_algorithms:
+            self._stabilize_capture(crypto_alg=crypto_alg)
```

Comment:
I think the stabilization can be factored out. We need to do it after a new firmware is flashed, after a power-cycle or things like this. But between datasets, this should be fine.